### DISCUSSION TITLE
AutoPID and Home Assistant Auto Discovery: Retain MQTT Messages

### DISCUSSION THREAD

**Discussion by iamthe1st** [meatpiHQ/wican-fw#198] [Ideas]
Hello,
I would like to thank @meatpiHQ for the great work and the AutoPID Feature!
I updated my WICAN, enabled the AutoPID Feature and the sensors popped up in Home Assistant flawlessly!!! 
Done in less then 20 Minutes - awesome!
I have a Kia Niro EV SG2 (Second Gen) and the KONA profile worked perfectly!

I have one minor Feature Request:
- change MQTT data values to the state "Retained" so that they are not lost after Home Assistant restart

@meatpiHQ i would like to buy you a drink/coffee - can you pleas post your paypal account for a donation?

**Comment by meatpiHQ** [comment]
@iamthe1st I'm always glad to hear something is actually working! Thank you! 

> change MQTT data values to the state "Retained" so that they are not lost after Home Assistant restart

Will be added in the next release.

> i would like to buy you a drink/coffee - can you pleas post your paypal account for a donation?

You don't have to, but thank you for the energy boost :)
https://github.com/sponsors/meatpiHQ?frequency=one-time&sponsor=meatpiHQ

**Comment by iamthe1st** [comment]
Hi,
i discovered two little problems after 2 day of usage:

First thing: all values apper "online" in MQTT Explorer, although the the wican is Offline (See MQTT topic \can\status = "offline" )
![image](https://github.com/user-attachments/assets/587cdb4d-4020-4d7f-b2a4-74de5cb2670f)

second:
in the middle of the night, the KIA NIRO EV seems to activate 12V for a short period of time. If this happens the the values obtained over AutoPID jump between the actual value and 0.5 %

This is identical for all AutoPID values.
Although the car is "off" the last value for most AutoPID "Sensors" in Home Assistant is 0.5% , the correct value or some other strange value
![image](https://github.com/user-attachments/assets/3d4191b6-543c-4481-b18b-20d502b95314)

![image](https://github.com/user-attachments/assets/06f06170-03a2-4290-94c2-f0f2b8d6c71c)


Maybe the strange value is coming from an CAN-Timeout?

**Reply by meatpiHQ** [reply]
> First thing: all values apper "online" in MQTT Explorer, although the the wican is Offline (See MQTT topic \can\status = "offline" )

I'm aware of this issue, and will be addressed in the next release 

> This is identical for all AutoPID values.
Although the car is "off" the last value for most AutoPID "Sensors" in Home Assistant is 0.5% , the correct value or some other strange value

Sorry, I don't fully understand the isse. The car is on the charger when it activates the 12V charging, and WiCAN wakeup? And it starts sending wrong values?

**Comment by iamthe1st** [comment]
Thank you for your quick reply!
Thanks for addressing this!

To the second topic:
In this case the car was parked in the garage with the WiCan connected inside the car.
The KIA NIRO EV switches off 12V on the OBD-Port when the car is locked, and is not charging.
Yesterday night the charger was not connected, but it seems that the NIRO wakes up at 21:00 and i guess is looking for OTA updates for the car, there it switches on the 12V on the OBD-Port.

But then the WiCAN sends strange values -- on the SOC values it jumps between the real value and 0.5% 
On the SOH value it jumps between 100% and 5.402,3%

**But** i had the same issue, when i was sending the RAW CAN-Commands over MQTT-publish with home assistant and "decrypting" them with the PID-Filter in the WICAN.

So **maybe it is just an special issue with my car.**


I introduced an templated sensor, which filters out all SOC values below 2%, i guess this will fix the issue for me.
Maybe this will also help others:


configuration.yaml in Home Assistant:

```
template:
  - sensor:
      - name: "niro_soc_filtered"
        state: >
          {% if states('sensor.soc_d') | float > 2 %}
           {{ states('sensor.soc_d') }}
          {% else %}
           {{ states('sensor.niro_soc_filtered') | float }}
          {% endif %}
        icon: mdi:battery
        unit_of_measurement: '%'
```


**sensor.soc_d** is the values coming from WICAN over AutoPID
**sensor.niro_soc_filtered** is the filtered sensor.
The value from sensor.soc_d is only copied over to sensor.niro_soc_filtered, when it is above 2%

**Reply by iamthe1st** [reply]
I left MQTT Explorer running -- this is the JSON Content, if the values are "wrong" 

Here SOC_BMS is wrong, Max and Min_CELL_V is also wrong - also AUX_BAT with 21.4V is wrong.

{
  "SOC_BMS": 0.5,
  "Max_REGEN": 112.65,
  "Max_Power": 3,
  "Batt_Current": 5911.1,
  "HV_Volts": 616.9,
  "Batt_MaxT": 24,
  "Batt_MinT": 25,
  "Batt_Temp_1": 53,
  "Batt_Temp_2": 201,
  "Batt_Temp_3": 71,
  "Batt_Temp_4": 24,
  "Batt_Temp_5": 0,
  "Batt_InletT": 100,
  "Max_Cell_V": 0.52,
  "Max_Cell_V_No": 0,
  "Min_Cell_V": 0.06,
  "Min_Cell_V_No": 232,
  "Aux_Batt_Volts": 21.4,
  "SOH": 100,
  "SOC_D": 86
}


Here a screenshot of the jumping values:

![image](https://github.com/user-attachments/assets/df039f9e-612d-4088-a4b4-044c1428808a)


**But again - it could also be my car.** 
**With the filtered sensor in home assistant, i sorted it out already, and have a working solution.**

**Reply by meatpiHQ** [reply]
I don't think it's an issue specific to your car, I think there some extra messages coming on the bus and can be filtered by using a proper initialisation string.

Can you please export logs from car scanner?

**Comment by iamthe1st** [comment]
Hi, today i let MQQT Explorer run while the car was charging and saw this.
Maybe the timeout is an explanation for the jumping values :-)

![image](https://github.com/user-attachments/assets/3ecb2506-4a42-4a2e-9771-51838bf7659d)

**Reply by meatpiHQ** [reply]
@iamthe1st i'm might be that the ecu went to sleep that's why it timed out, but in this case it just won't send anything.

**Comment by jmalm** [comment]
> Hello, I would like to thank @meatpiHQ for the great work and the AutoPID Feature! I updated my WICAN, enabled the AutoPID Feature and the sensors popped up in Home Assistant flawlessly!!! Done in less then 20 Minutes - awesome!

I assume that is the feature under Automate where, if you enable "Vehicle Specific", and upload a "Vehicle Profiles" JSON file, you can select a (more or less) specific car model to publish various informations for over MQTT. Correct?

> I have a Kia Niro EV SG2 (Second Gen) and the KONA profile worked perfectly!

Is that the "Hyundai: Ioniq5/Ioniq6/Kona" profile from https://github.com/meatpiHQ/wican-fw/blob/main/vehicle_profiles.json?
I have a Kia e-Niro (2019, i.e. first generation). Do you know if the same profile should work with my car?
I can't get it to work with the following configuration:

![image](https://github.com/user-attachments/assets/1f243e53-56b2-40b8-877e-0d2539781231)

I've also tried to use "test" as topic - no change. I never get any data other than status (on the topic that I've configured for that in Settings).

Any ideas?

**Reply by iamthe1st** [reply]
Yes, this profile should work.
Before buying the Niro I also test drove the gen1 Niro and the same carscanner profile worked on both Niro Gen 1 and 2.

**Reply by jmalm** [reply]
Hmm... I wonder what I might be doing wrong. If I don't enter a topic, which topic will be used? I would have expected messages to be published on homeassistant/sensor/something.

**Reply by iamthe1st** [reply]
For all MQTT devices to work, you have to install the mosquitto addon in Home Assistant.
In your wican you have to type your newly created mosquitto server as base for the topic

Once this is done, you can go to devices - > mqtt in home assistant.
There you will find the "sensors" from your wican

**Reply by iamthe1st** [reply]
Search for an guide to install and set up mosquitto in home assistant

**Reply by jmalm** [reply]
Thanks for the suggestion. I have mosquitto running a number of integrations for home assistant already.

With the wican unit connected in my car, I'm using mqtt explorer to inspect the published messages as they come in on the bus, but the only message that comes in from the wican unit with this setup is the status message, just after boot.

**Reply by jmalm** [reply]
I found the issue. I hadn't set the protocol to AutoPID. It's working correctly now. 

Thanks for the help!
